pkg/tfbridge/detailed_diff.go

```diff
@@ -0,0 +1,484 @@
+	if isPresent(old, oldOk) && old.IsObject() {
+		oldObj = old.ObjectValue()
+	}
+	if isPresent(new, newOk) && new.IsObject() {
+		newObj = new.ObjectValue()
+	}
+
+	keys := make(map[resource.PropertyKey]struct{})
+	for k := range oldObj {
+		keys[k] = struct{}{}
+	}
+	for k := range newObj {
+		keys[k] = struct{}{}
+	}
+
+	for k := range keys {
+		key := getSubPath(key, k)
+		oldVal, oldOk := oldObj[k]
+		newVal, newOk := newObj[k]
+		_, etf, eps := getInfoFromPulumiName(k, etf, eps)
+
+		propDiff := makePropDiff(ctx, key, etf, eps, oldVal, newVal, oldOk, newOk)
+
+		for subKey, subDiff := range propDiff {
+			diff[subKey] = subDiff
+		}
+	}
+
+	simplerDiff, err := simplifyDiff(diff, key, old, new, oldOk, newOk, etf, nil)
+	if err == nil {
+		return simplerDiff
+	}
+
+	return diff
+}
+
+func makeElemDiff(
+	ctx context.Context,
+	key resource.PropertyKey,
+	etf interface{},
+	eps *SchemaInfo,
+	parentForceNew bool,
+	old, new resource.PropertyValue,
+	oldOk, newOk bool,
+) map[string]*pulumirpc.PropertyDiff {
+	diff := make(map[string]*pulumirpc.PropertyDiff)
+	if _, ok := etf.(shim.Resource); ok {
+		fields := map[string]*SchemaInfo{}
+		if eps != nil {
+			fields = eps.Fields
+		}
+		etfSch := etf.(shim.Resource).Schema()
+		d := makeObjectDiff(
+			ctx,
+			key,
+			etfSch,
+			fields,
+			old,
+			new,
+			oldOk,
+			newOk,
+		)
+		for subKey, subDiff := range d {
+			diff[subKey] = subDiff
+		}
+	} else if _, ok := etf.(shim.Schema); ok {
+		d := makePropDiff(
+			ctx,
+			key,
+			etf.(shim.Schema),
+			eps,
+			old,
+			new,
+			oldOk,
+			newOk,
+		)
+		for subKey, subDiff := range d {
+			if parentForceNew {
+				subDiff = promoteToReplace(subDiff)
+			}
+			diff[subKey] = subDiff
+		}
+	} else {
+		d := makePropDiff(ctx, key, nil, eps.Elem, old, new, oldOk, newOk)
+		for subKey, subDiff := range d {
+			if parentForceNew {
+				subDiff = promoteToReplace(subDiff)
+			}
+			diff[subKey] = subDiff
+		}
+	}
+
+	simplerDiff, err := simplifyDiff(diff, key, old, new, oldOk, newOk, etf, eps)
+	if err == nil {
+		return simplerDiff
+	}
+
+	return diff
+}
+
+func makeListDiff(
+	ctx context.Context,
+	key resource.PropertyKey,
+	etf shim.Schema,
+	eps *info.Schema,
+	old, new resource.PropertyValue,
+	oldOk, newOk bool,
+) map[string]*pulumirpc.PropertyDiff {
+	diff := make(map[string]*pulumirpc.PropertyDiff)
+	oldList := []resource.PropertyValue{}
+	newList := []resource.PropertyValue{}
+	if isPresent(old, oldOk) && old.IsArray() {
+		oldList = old.ArrayValue()
+	}
+	if isPresent(new, newOk) && new.IsArray() {
+		newList = new.ArrayValue()
+	}
+
+	// naive diffing of lists
+	// TODO[pulumi/pulumi-terraform-bridge#2295]: implement a more sophisticated diffing algorithm
+	// investigate how this interacts with force new - is identity preserved or just order
+	collectionForceNew := isForceNew(etf, eps)
+	longerLen := max(len(oldList), len(newList))
+	for i := 0; i < longerLen; i++ {
+		elemKey := string(key) + "[" + fmt.Sprintf("%d", i) + "]"
```

Comment:
This is really sketchy from type-safety POV and begs the question what does makeElemDiff etc do with the key parameter? Are we building up the nested paths for the `map[string]*pulumirpc.PropertyDiff` keys? Perhaps throw some types to make it clear?